src/services/GraphsDFG.jl

```diff
@@ -0,0 +1,432 @@
+    if haskey(dfg.labelDict, factor.label)
+        error("Factor '$(factor.label)' already exists in the factor graph")
+    end
+    for v in variables
+        if !(v.label in keys(dfg.labelDict))
+            error("Variable '$(v.label)' not found in graph when creating Factor '$(factor.label)'")
+        end
+    end
+    dfg.nodeCounter += 1
+    factor._internalId = dfg.nodeCounter
+    fNode = GraphsNode(dfg.nodeCounter, factor)
+    f = Graphs.add_vertex!(dfg.g, fNode)
+    # Add index
+    push!(dfg.labelDict, factor.label=>factor._internalId)
+    # Add the edges...
+    for variable in variables
+        v = dfg.g.vertices[variable._internalId]
+        edge = Graphs.make_edge(dfg.g, f, v)
+        Graphs.add_edge!(dfg.g, edge)
+    end
+    return true
+end
+
+"""
+    $(SIGNATURES)
+Get a DFGVariable from a DFG using its underlying integer ID.
+"""
+function getVariable(dfg::GraphsDFG, variableId::Int64)::DFGVariable
+    @warn "This may be slow, rather use by getVariable(dfg, label)"
+    #TODO: This may be slow (O(n)), can we make it better?
+    if !(variableId in values(dfg.labelDict))
+        error("Variable ID '$(variableId)' does not exist in the factor graph")
+    end
+    return dfg.g.vertices[variableId].dfgNode
+end
+
+"""
+    $(SIGNATURES)
+Get a DFGVariable from a DFG using its label.
+"""
+function getVariable(dfg::GraphsDFG, label::Union{Symbol, String})::DFGVariable
+    if typeof(label) == String
+        label = Symbol(label)
+    end
+    if !haskey(dfg.labelDict, label)
+        error("Variable label '$(label)' does not exist in the factor graph")
+    end
+    return dfg.g.vertices[dfg.labelDict[label]].dfgNode
+end
+
+"""
+    $(SIGNATURES)
+Get a DFGFactor from a DFG using its underlying integer ID.
+"""
+function getFactor(dfg::GraphsDFG, factorId::Int64)::DFGFactor
+    @warn "This may be slow, rather use by getFactor(dfg, label)"
+    #TODO: This may be slow (O(n)), can we make it better?
+    if !(factorId in values(dfg.labelDict))
+        error("Factor ID '$(factorId)' does not exist in the factor graph")
+    end
+    return dfg.g.vertices[factorId].dfgNode
+end
+
+"""
+    $(SIGNATURES)
+Get a DFGFactor from a DFG using its label.
+"""
+function getFactor(dfg::GraphsDFG, label::Union{Symbol, String})::DFGFactor
+    if typeof(label) == String
+        label = Symbol(label)
+    end
+    if !haskey(dfg.labelDict, label)
+        error("Factor label '$(label)' does not exist in the factor graph")
+    end
+    return dfg.g.vertices[dfg.labelDict[label]].dfgNode
+end
+
+"""
+    $(SIGNATURES)
+Update a complete DFGVariable in the DFG.
+"""
+function updateVariable!(dfg::GraphsDFG, variable::DFGVariable)::DFGVariable
+    if !haskey(dfg.labelDict, variable.label)
+        error("Variable label '$(variable.label)' does not exist in the factor graph")
+    end
+    dfg.g.vertices[dfg.labelDict[variable.label]].dfgNode = variable
```

Comment:
I had cases where I ably want to update some of the fields, more a user thing and I’m not exactly sure of how the new use case will work, but just keep in mind till we have completed integration.

Reply:
Yep! That's part of the actual DFGVariable itself though, more on that when adding the fields into the variables/factors.